### PR TITLE
Improvements to toast context loading plus small fixes

### PR DESCRIPTION
**Pull request by tskisner**
- Add support for both collective and independent reading of data from the context.

- Add support for specifying the full detector selection dictionary.

- Add ability to apply preprocessing if a preprocessing config is specified.

- Avoid loading any data when distributing observations among process groups.

- In mlmapmaker wrapper, disable MPI in pixell if it is disabled in toast.  This prevents mpi4py imports on NERSC login nodes.

- Small fix in specifying workflow configs when running interactively.

### CODE REVIEW

File: sotodlib/toast/ops/load_context.py [hunk 1]
```diff
@@ -87,6 +100,8 @@ class LoadContext(Operator):
 
     bands = List(list(), help="Only load this list of band values")
 
+    dets_select = Dict(dict(), help="The full detector selection dictionary to use")
```
Comment (keskitalo) [inline sotodlib/toast/ops/load_context.py:103]: This seems a little cryptic to me. It is not a list of detectors but a dictionary. Is each key a name of a detector set?

File: sotodlib/toast/ops/load_context.py [hunk 2]
```diff
@@ -207,6 +228,16 @@ class LoadContext(Operator):
...
         super().__init__(**kwargs)
 
+    def _open_context(self):
+        if self.context is None:
+            return Context(self.context_file)
+        else:
+            return self.context
+
+    def _close_context(self, ctx):
+        if self.context is None:
```
Comment (keskitalo) [inline sotodlib/toast/ops/load_context.py:238]: Is this sufficient or should you set `self.context` to `None`?

File: sotodlib/toast/ops/load_context.py [hunk 3]
```diff
@@ -367,7 +464,7 @@ def _exec(self, data, detectors=None, **kwargs):
 
             focalplane = toast.instrument.Focalplane(
                 detector_data=det_props,
-                sample_rate=rate,
+                sample_rate=1.0 * u.Hz,
```
Comment (keskitalo) [inline sotodlib/toast/ops/load_context.py:467]: Is this a leftover debugging setting?

**Review by keskitalo — APPROVED** [review]
Looks good. Please have a look at the small questions.

File: sotodlib/toast/ops/load_context.py (hunk 1 above)
Comment (tskisner) [reply sotodlib/toast/ops/load_context.py:103]: This dictionary is passed [directly to the Context method](https://github.com/simonsobs/sotodlib/blob/8c47e388f65841e4bc7b5af16b320100324970f3/sotodlib/core/context.py#L223).  I'll add a note in the trait help.

File: sotodlib/toast/ops/load_context.py (hunk 2 above)
Comment (tskisner) [reply sotodlib/toast/ops/load_context.py:238]: There are 2 traits at play here, `context` and `context_file`.  The user specifies one but not both.  If `context` is set to a pre-existing context instance, then "open" just returns that, else it instantiates a context from `context_file` and returns.  Similarly, if `context` was given, we don't try to close it.  But if we created it then we delete it.

File: sotodlib/toast/ops/load_context.py (hunk 3 above)
Comment (tskisner) [reply sotodlib/toast/ops/load_context.py:467]: No, this is an optimization.  When we create the observation, we do not yet have timestamps (only the number of samples).  So we cannot compute the sample rate yet without reading through the entire dataset (which is what was happening before in the code).  Instead, we set the focalplane to a fake sample rate and later (after the timestamps are loaded) we go back and fix the sample rate.

File: sotodlib/toast/ops/load_context.py (hunk 1 above)
Comment (tskisner) [reply sotodlib/toast/ops/load_context.py]: Ok, trait help is reworded.

**Pull request merged**